### ISSUE TITLE
Add support for creating custom modules

### ISSUE THREAD

**Issue by jumale** [wtfutil/wtf#322]
#### What problem does this solve?
It is a powerful tool and provides already many ready-to-use widgets.
It also has the great modular architecture which suppose to help for creating more and more modules... but unfortunately the are some decisions which make it not that easy.
I'm going to use wtf as a base platform for developing my own dashboard on table-based widgets, therefore I've cloned this project and I'm going to adjust it in the way that it has more flexibility enough to inject my modules.

Here are some problems which I faced so far:

**1.** it's not documented how to create custom modules

**2.** `main.go` takes too much responsibility for bootstrapping an application (it's hard to create a new app with another widgets setup)

**3.** there is only one base widget `TextWidget` (implemented with `tview.TextView`) which is used by most of widgets, while `tview` provides much more primitives able to draw. Personally I need a table-based widget, but it's not possible to do it out from the box, because `Wtfable` interface is coupled to `tview.TextView`

**4.** widgets have too much knowledge about config paths by using global Config object and full config path prefixes (e.g `wtf.mods.%s.enabled`). This is not something what blocks me at the moment, but it might be painful in future, if those paths gonna change.

#### How do envision it working?

To fix those problems, I'm doing the next changes:
**4.** (nice to have) get rid of global config. Instead, make widgets configurable by a method, so that they receive only their own config, which they can process as they wish. This will make widgets more independent from the root project and its global variables, and make it easier to develop widgets in outer packages.

**3.** Adjust `Wtfable` interface and all related dependencies to allow any `tview.Primitive`-based widgets, but not only `tview.TextView`. Add another base widget `wtf/table_widget.go` which is based on `tview.Table` primitive.

**2.** Extract most of logic in the main.go into an app object. It should be possible to do something line this:

```go
// main.go
app := wtf.NewApp()
app.LoadConfig("my_config.yml")

app.AddWidget(github.Widget{})
app.AddWidget(clocks.Widget{})

app.Run()
```
Then we can easily create more than one setup of wtf app

**1.** If all previous steps are successfully done and everything works as expected - I would make a PR to this project to make it accessible for everybody. If it's approved and going to be merged, then we can add an official documentation about "How to write custom modules".

Please let me know if you're interested in those changes, then I'll spend more time to be sure that the new approach is flexible enough to be used by anyone else (otherwise I'll only adjust the WTF project for my needs and will be using it from my fork).

**Comment by senorprogrammer** [comment]
@jumale I completely agree with everything you wrote above. You've done an awesome job of enumerating the fundamental design deficiencies and these (amongst other reasons) are why I've kept the release version number so low. Any movement you make in this direction is definitely welcome and greatly appreciated.

**Comment by jumale** [comment]
@senorprogrammer I was on vacation for 2 weeks. Now I'm back and will keep working on this. Unfortunately I should admit that the resulting PR will contains a lot of lines of changes (because of migrating the config from globals to injected version, as well as changing the way how widgets get initialised). But I will create at the end a migration doc with explanations for common changes - it should make the review easier.

**Comment by senorprogrammer** [comment]
Looking forward to checking it out! If you could, make any PRs as atomic as possible. There's a lot of excellent concepts mentioned above, if they could be broken down logically I'd appreciate that. Thanks for making this effort.

**Comment by jumale** [comment]
hey @senorprogrammer 
It took more time, than I expected - I was quite busy with my main job.
At some moment I have finished refactoring the part, which I needed for my custom module, but then it went a bit crazy and I changed another parts which were not actually blocking me, but just for improving the structure and extensibility.
I understand that it does not look now like a small PR, which can be easily reviewed and merged... But I'm open for discussions. Please check [the readme](https://github.com/jumale/wtf/blob/new-app-arch/REFACTORING.md) where I have documented most of the changes. All the changes are pushed to this [separated branch of my fork](https://github.com/jumale/wtf/tree/new-app-arch).

**Note:** I did not migrate all the widgets to the new format. Only a the widgets which are used in the sample config, to show an example of how the app works. Plus the `github` widget, because I actually use it.

So, when you have time, check the changes, let me know what you think about all this... and then we can think about what to do with that code and which parts of it can go to this origin repository.

**Comment by senorprogrammer** [comment]
Thanks @jumale, I'll try to dig into this this weekend. It looks like you've done an amazing amount of work!

**Comment by senorprogrammer** [comment]
I haven't had a chance to dig into your changes and improvements yet, however I have done the following:

* moved WTF to an organization called [WTFUtil](https://github.com/wtfutil) to make it not specifically my project anymore (at this point others have contributed to it more than I have)
* added you as a collaborator on this repo and a member on the organization, if you like, which should give you the ability to add change without having to wait on me

Philosophically, I think we're in line with how things should improve. I only ask a few things:

* no breaking changes as far as the user is concerned. Each version should continue to work for them without them having to do anything manual to make it so
* follow semver and document all changes

**Comment by jumale** [comment]
@senorprogrammer thank you for your trust 👍🏻 I will be happy to contribute to the project.
I will try to find a point from where to start introducing the changes. But probably I won't have much time this month, because of a pile of tasks on my main job, waiting for finishing before the end of the year. I will probably prepare a migration plan, for non-breaking changes, and then start implementing it in January.

**Comment by senorprogrammer** [comment]
@jumale I hear you. I'm in a similar boat, having started a new job a couple months ago. Finding spare time is definitely tough. I really appreciate all the effort and work you two have put in, especially you @anandsudhir, for picking up when I've dropped the ball.

**Comment by senorprogrammer** [comment]
An update:

#2, from up above, has been addressed. `main` is much smaller now.

#4 has been partially addressed from the module perspective. Config use is now separate from the underlying implementation.

#1 is still valid and I've been holding off on that until it's much easier to create a new module. It should be simple, allowing the implementor to focus on the unique details of their service. Right now a big chunk is still dealing with boilerplate.

#3 stay tuned.
